File: facebook-core/src/main/java/com/facebook/appevents/iap/InAppPurchaseAutoLogger.java
/*
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook.appevents.iap;

import android.content.Context;
import androidx.annotation.RestrictTo;
import com.facebook.internal.instrument.crashshield.AutoHandleExceptions;

@AutoHandleExceptions
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
public class InAppPurchaseAutoLogger {
  private static final String BILLING_CLIENT_PURCHASE_NAME =
      "com.android.billingclient.api.Purchase";

  public static void startIapLogging(Context context) {
    // check if the app has IAP with Billing Lib
    if (InAppPurchaseUtils.getClass(BILLING_CLIENT_PURCHASE_NAME) == null) {
      return;
    }
    final InAppPurchaseBillingClientWrapper billingClientWrapper =
        InAppPurchaseBillingClientWrapper.getOrCreateInstance(context);
    if (billingClientWrapper == null) {
      return;
    }

    if (InAppPurchaseBillingClientWrapper.isServiceConnected.get()) {
      if (InAppPurchaseLoggerManager.eligibleQueryPurchaseHistory()) {
        billingClientWrapper.queryPurchaseHistory(
            "inapp",
            new Runnable() {
              @Override
              public void run() {
                logPurchase();
              }
            });
      } else {
        billingClientWrapper.queryPurchase(
            "inapp",
            new Runnable() {
              @Override
              public void run() {
                logPurchase();
              }
            });
      }
    }
  }

  private static void logPurchase() {
    InAppPurchaseLoggerManager.filterPurchaseLogging(
        InAppPurchaseBillingClientWrapper.purchaseDetailsMap,
        InAppPurchaseBillingClientWrapper.skuDetailsMap);
    InAppPurchaseBillingClientWrapper.purchaseDetailsMap.clear();
  }
}

File: facebook-core/src/main/java/com/facebook/appevents/iap/InAppPurchaseAutoLogger.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.appevents.iap

import com.facebook.appevents.iap.InAppPurchaseUtils.BillingClientVersion.V2_V4
import com.facebook.appevents.iap.InAppPurchaseUtils.BillingClientVersion.V5_V7
import com.facebook.appevents.iap.InAppPurchaseUtils.IAPProductType.INAPP
import android.content.Context
import androidx.annotation.RestrictTo
import com.facebook.appevents.iap.InAppPurchaseLoggerManager.getIsFirstAppLaunch
import com.facebook.appevents.iap.InAppPurchaseLoggerManager.setAppHasBeenLaunched
import com.facebook.appevents.iap.InAppPurchaseUtils.IAPProductType.SUBS
import com.facebook.appevents.integrity.ProtectedModeManager
import com.facebook.internal.FeatureManager
import com.facebook.internal.FeatureManager.isEnabled
import com.facebook.internal.instrument.crashshield.AutoHandleExceptions
import java.util.concurrent.atomic.AtomicBoolean

@AutoHandleExceptions
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
object InAppPurchaseAutoLogger {
    val failedToCreateWrapper = AtomicBoolean(false)

    @JvmStatic
    fun startIapLogging(
        context: Context,
        billingClientVersion: InAppPurchaseUtils.BillingClientVersion
    ) {
        // Check if we have previously tried and failed to create a billing client wrapper
        if (failedToCreateWrapper.get()) {
            return
        }
        var billingClientWrapper: InAppPurchaseBillingClientWrapper? = null
        if (billingClientVersion == V2_V4) {
            billingClientWrapper =
                InAppPurchaseBillingClientWrapperV2V4.getOrCreateInstance(context)
        } else if (billingClientVersion == V5_V7) {
            billingClientWrapper =
                InAppPurchaseBillingClientWrapperV5V7.getOrCreateInstance(context)
        }
        if (billingClientWrapper == null) {
            failedToCreateWrapper.set(true)
            return
        }

        if (isEnabled(FeatureManager.Feature.AndroidIAPSubscriptionAutoLogging)
            && (!ProtectedModeManager.isEnabled() || billingClientVersion == V2_V4)
        ) {
            billingClientWrapper.queryPurchaseHistory(INAPP) {
                billingClientWrapper.queryPurchaseHistory(SUBS) {
                    logPurchase(billingClientVersion, context.packageName)
                }
            }
        } else {
            billingClientWrapper.queryPurchaseHistory(INAPP) {
                logPurchase(billingClientVersion, context.packageName)
            }
        }
    }

    private fun logPurchase(
        billingClientVersion: InAppPurchaseUtils.BillingClientVersion,
        packageName: String
    ) {
        val isFirstAppLaunch = getIsFirstAppLaunch()
        if (billingClientVersion == V2_V4) {
            InAppPurchaseLoggerManager.filterPurchaseLogging(
                InAppPurchaseBillingClientWrapperV2V4.iapPurchaseDetailsMap,
                InAppPurchaseBillingClientWrapperV2V4.skuDetailsMap,
                false,
                packageName,
                billingClientVersion,
                isFirstAppLaunch
            )
            InAppPurchaseLoggerManager.filterPurchaseLogging(
                InAppPurchaseBillingClientWrapperV2V4.subsPurchaseDetailsMap,
                InAppPurchaseBillingClientWrapperV2V4.skuDetailsMap,
                true,
                packageName,
                billingClientVersion,
                isFirstAppLaunch
            )
            InAppPurchaseBillingClientWrapperV2V4.iapPurchaseDetailsMap.clear()
            InAppPurchaseBillingClientWrapperV2V4.subsPurchaseDetailsMap.clear()
        } else {
            InAppPurchaseLoggerManager.filterPurchaseLogging(
                InAppPurchaseBillingClientWrapperV5V7.iapPurchaseDetailsMap,
                InAppPurchaseBillingClientWrapperV5V7.productDetailsMap,
                false,
                packageName,
                billingClientVersion,
                isFirstAppLaunch
            )
            InAppPurchaseLoggerManager.filterPurchaseLogging(
                InAppPurchaseBillingClientWrapperV5V7.subsPurchaseDetailsMap,
                InAppPurchaseBillingClientWrapperV5V7.productDetailsMap,
                true,
                packageName,
                billingClientVersion,
                isFirstAppLaunch
            )
            InAppPurchaseBillingClientWrapperV5V7.iapPurchaseDetailsMap.clear()
            InAppPurchaseBillingClientWrapperV5V7.subsPurchaseDetailsMap.clear()
        }
        if (isFirstAppLaunch) {
            setAppHasBeenLaunched()
        }
    }
}
